plugins/modules/dcnm_service_policy.py

```diff
@@ -0,0 +1,1962 @@
+        input. This routine updates self.sp_info with validated playbook information related to query
+        state.
+
+        Parameters:
+            cfg (dict): The config from playbook
+
+        Returns:
+           None
+        """
+
+        sp_query_spec = dict(
+            node_name=dict(required=True, type="str"),
+            name=dict(required=False, type="str", default="None"),
+        )
+
+        sp_info, invalid_params = validate_list_of_dicts(cfg, sp_query_spec)
+        if invalid_params:
+            mesg = "Invalid parameters in playbook: {}".format(
+                "while processing Service Policy -  "
+                + cfg[0]["name"]
+                + ", "
+                + "\n".join(invalid_params)
+            )
+            self.module.fail_json(msg=mesg)
+
+        self.sp_info.extend(sp_info)
+
+    def dcnm_sp_get_service_node_type(self, sp):
+
+        path = (
+            "/appcenter/Cisco/elasticservice/elasticservice-api/fabrics/"
+            + self.module.params["attached_fabric"]
+            + "/service-nodes/"
+            + sp["node_name"]
+        )
+
+        retries = 0
+        while retries <= 30:
+
+            retries += 1
+            resp = dcnm_send(self.module, "GET", path)
+
+            if resp and (resp["RETURN_CODE"] == 200) and resp["DATA"]:
+                return resp["DATA"].get("type", None)
+            else:
+                self.dcnm_sp_check_unauthorized_error_in_resp(resp)
+                time.sleep(10)
+                continue
+        self.changed_dict[0]["debugs"].append({"GET_SNODE_TYPE": resp})
+        return "Firewall"
+
+    def dcnm_sp_get_sp_payload(self, sp):
+
+        """
+        This routine builds the complete service policy payload based on the information in self.want
+
+        Parameters:
+            sp (dict): Route peering information
+
+        Returns:
+            sp_payload (dict): SP payload information populated with appropriate data from playbook config
+        """
+
+        node_type = self.dcnm_sp_get_service_node_type(sp)
+
+        sp_payload = {}
+
+        # Populate the common information
+
+        sp_payload["serviceNodeName"] = sp["node_name"]
+        sp_payload["attachedFabricName"] = self.module.params["fabric"]
+        sp_payload["serviceNodeType"] = node_type
+        sp_payload["fabricName"] = self.module.params["attached_fabric"]
+        sp_payload["enabled"] = str(self.attach).lower()
+
+        sp_payload["policyName"] = sp["name"]
+        sp_payload["peeringName"] = sp["rp_name"]
+        sp_payload["sourceVrfName"] = sp["src_vrf"]
+        sp_payload["destinationVrfName"] = sp["dest_vrf"]
+        sp_payload["sourceNetwork"] = sp["src_network"]
+        sp_payload["destinationNetwork"] = sp["dest_network"]
+        sp_payload["reverseEnabled"] = sp["reverse"]
+        sp_payload["policyTemplateName"] = "service_pbr"
+        sp_payload["nextHopIp"] = sp["next_hop"]
+        if sp["reverse"]:
```

Comment:
Is `sp` a dict?  If so should you use `sp.get("reverse")` just to be safe?

Reply:
We can. But sp["reverse"] is defaulted to True in Validate function if it is not given in the playbook. So we are safe.